File: 11-integrating-with-function-calling/typescript/mcp-roundtable-example.ts
#!/usr/bin/env node
/**
 * Model Context Protocol (MCP) Educational Example with Roundtable - TypeScript
 * ==========================================================================
 *
 * This TypeScript example demonstrates how to use the Roundtable MCP Server
 * for educational purposes, building on the function calling concepts from
 * the main lesson.
 *
 * Prerequisites:
 * - npm install -g @roundtable/mcp-server (coming soon)
 * - OR: pip install roundtable-ai (current method)
 * - At least one supported AI assistant installed
 *
 * Learning Objectives:
 * - Understand how MCP extends function calling in TypeScript/JavaScript
 * - See unified AI assistant management in web development contexts
 * - Learn to build educational web applications with multiple AI tools
 * - Integrate MCP with modern JavaScript frameworks
 */

import { spawn, exec } from 'child_process';
import { promisify } from 'util';
import * as fs from 'fs/promises';
import * as path from 'path';

const execAsync = promisify(exec);

interface DemoQuestion {
    question: string;
    category: string;
    bestAI: string;
    webContext: string;
}

interface MCPConfig {
    mcpServers: {
        [key: string]: {
            command: string;
            env?: Record<string, string>;
            args?: string[];
        };
    };
}

class EducationalMCPDemoTS {
    private availableAssistants: string[] = [];
    private demoQuestions: DemoQuestion[] = [
        {
            question: "Explain what a REST API is and how to use it in JavaScript",
            category: "web-concepts",
            bestAI: "gemini",
            webContext: "Essential for frontend-backend communication"
        },
        {
            question: "Create a React component for a todo list",
            category: "frontend-coding",
            bestAI: "codex",
            webContext: "Clean, reusable component architecture"
        },
        {
            question: "Review this Express.js route for security issues",
            category: "backend-review",
            bestAI: "claude",
            webContext: "Security and best practices analysis"
        },
        {
            question: "Debug this async/await function step by step",
            category: "debugging",
            bestAI: "cursor",
            webContext: "Interactive debugging for async code"
        }
    ];

    async checkRoundtableInstallation(): Promise<boolean> {
        try {
            const { stdout, stderr } = await execAsync('roundtable-ai --help');
            return !stderr && stdout.includes('roundtable');
        } catch (error) {
            return false;
        }
    }

    async discoverAvailableAssistants(): Promise<string[]> {
        try {
            const { stdout } = await execAsync('roundtable-ai --check');

            const available: string[] = [];
            const assistants = ['claude', 'codex', 'gemini', 'cursor'];

            for (const assistant of assistants) {
                if (stdout.toLowerCase().includes(assistant) &&
                    (stdout.includes('✓') || stdout.includes('available'))) {
                    available.push(assistant);
                }
            }

            return [...new Set(available)]; // Remove duplicates
        } catch (error) {
            console.log('⚠️  Discovery timeout or error - this is normal for first run');
            return [];
        }
    }

    async demonstrateMCPConcepts(): Promise<void> {
        console.log('🎓 Model Context Protocol (MCP) Educational Demo - TypeScript Edition');
        console.log('='.repeat(70));

        // 1. Installation Check
        console.log('\n1️⃣ Checking Roundtable Installation...');
        const isInstalled = await this.checkRoundtableInstallation();

        if (!isInstalled) {
            console.log('❌ Roundtable not installed');
            console.log('📋 Install with: pip install roundtable-ai');
            console.log('📋 (npm package coming soon!)');
            return;
        }
        console.log('✅ Roundtable is installed!');

        // 2. Tool Discovery
        console.log('\n2️⃣ Discovering Available AI Assistants...');
        this.availableAssistants = await this.discoverAvailableAssistants();

        if (this.availableAssistants.length === 0) {
            console.log('⚠️  No AI assistants detected');
            console.log('📋 Install at least one: claude, codex, gemini, or cursor');
        } else {
            console.log(`✅ Found ${this.availableAssistants.length} assistants:`);
            this.availableAssistants.forEach(assistant => {
                console.log(`   • ${assistant}`);
            });
        }

        // 3. Web Development Use Cases
        console.log('\n3️⃣ Web Development Use Cases for MCP:');
        this.demonstrateWebDevelopmentScenarios();

        // 4. Framework Integration Examples
        console.log('\n4️⃣ Framework Integration Examples:');
        this.showFrameworkIntegrations();
    }

    demonstrateWebDevelopmentScenarios(): void {
        const scenarios = {
            "Full-Stack Learning": {
                description: "Use different AIs for frontend, backend, and deployment",
                example: "React (Cursor) → Express API (Codex) → Review (Claude) → Deploy (Gemini)"
            },
            "Code Review Workflow": {
                description: "Multi-perspective code review for web applications",
                example: "Security review (Claude), Performance (Codex), UX (Gemini), Debugging (Cursor)"
            },
            "Educational Tutorials": {
                description: "Create comprehensive web development tutorials",
                example: "Concept explanation (Gemini) → Code examples (Codex) → Best practices (Claude)"
            },
            "Real-time Development": {
                description: "Switch AIs during development based on current needs",
                example: "Stuck on CSS? → Cursor. Need API design? → Codex. Want explanation? → Gemini"
            }
        };

        Object.entries(scenarios).forEach(([scenario, details]) => {
            console.log(`\n📚 ${scenario}:`);
            console.log(`   • ${details.description}`);
            console.log(`   • Example: ${details.example}`);
        });
    }

    showFrameworkIntegrations(): void {
        console.log('\n🎯 JavaScript Framework Examples:');

        this.demoQuestions.forEach((demo, index) => {
            console.log(`\n🔧 Example ${index + 1}: ${demo.category}`);
            console.log(`   Question: ${demo.question}`);
            console.log(`   Best AI: ${demo.bestAI}`);
            console.log(`   Web Context: ${demo.webContext}`);
            console.log(`   Why: ${this.getAIStrengthForWeb(demo.bestAI)}`);
        });
    }

    getAIStrengthForWeb(aiName: string): string {
        const strengths = {
            "claude": "Excellent for security review, code quality, and web best practices",
            "codex": "Generates clean React/Node.js code with modern patterns",
            "gemini": "Great for explaining web concepts, API design, and architecture",
            "cursor": "Interactive debugging and step-by-step web development guidance"
        };
        return strengths[aiName as keyof typeof strengths] || "General purpose web development assistant";
    }

    async createWebDevelopmentMCPConfigs(): Promise<void> {
        console.log('\n📝 Web Development MCP Configurations:');
        console.log('='.repeat(45));

        const configs = {
            "Next.js Project": {
                filePath: ".cursor/mcp.json",
                config: {
                    mcpServers: {
                        "roundtable-ai": {
                            command: "roundtable-ai",
                            env: {
                                CLI_MCP_SUBAGENTS: "codex,claude,cursor,gemini",
                                CLI_MCP_WORKING_DIR: process.cwd()
                            }
                        }
                    }
                }
            },
            "VS Code Web Project": {
                filePath: ".vscode/settings.json",
                config: {
                    "mcp.servers": {
                        "roundtable-ai": {
                            command: "roundtable-ai",
                            env: {
                                CLI_MCP_SUBAGENTS: "codex,claude,cursor,gemini",
                                CLI_MCP_WORKING_DIR: "${workspaceFolder}"
                            }
                        }
                    }
                }
            },
            "Package.json Scripts": {
                filePath: "package.json",
                config: {
                    scripts: {
                        "ai:check": "roundtable-ai --check",
                        "ai:start": "roundtable-ai",
                        "ai:codex": "roundtable-ai --agents codex",
                        "ai:claude": "roundtable-ai --agents claude",
                        "ai:all": "roundtable-ai --agents codex,claude,cursor,gemini"
                    }
                }
            }
        };

        for (const [platform, details] of Object.entries(configs)) {
            console.log(`\n🔧 ${platform}:`);
            console.log(`   File: ${details.filePath}`);
            console.log(`   Configuration:`);
            console.log(JSON.stringify(details.config, null, 6));
        }
    }

    async generateWebProjectExample(): Promise<void> {
        console.log('\n🚀 Web Project Example Using Multiple AIs:');
        console.log('='.repeat(45));

        const projectFlow = [
            {
                phase: "Project Planning",
                ai: "gemini",
                task: "Research best practices for todo app architecture",
                code: `// Ask Gemini: "What's the best architecture for a React todo app with Node.js backend?"`
            },
            {
                phase: "Frontend Component",
                ai: "codex",
                task: "Generate clean React component code",
                code: `
// Generated by Codex
import React, { useState } from 'react';

interface Todo {
  id: string;
  text: string;
  completed: boolean;
}

const TodoList: React.FC = () => {
  const [todos, setTodos] = useState<Todo[]>([]);
  const [inputText, setInputText] = useState('');

  const addTodo = () => {
    if (inputText.trim()) {
      const newTodo: Todo = {
        id: Date.now().toString(),
        text: inputText,
        completed: false
      };
      setTodos([...todos, newTodo]);
      setInputText('');
    }
  };

  return (
    <div className="todo-app">
      <input
        value={inputText}
        onChange={(e) => setInputText(e.target.value)}
        onKeyPress={(e) => e.key === 'Enter' && addTodo()}
        placeholder="Add a new todo..."
      />
      <button onClick={addTodo}>Add Todo</button>
      <ul>
        {todos.map(todo => (
          <li key={todo.id} className={todo.completed ? 'completed' : ''}>
            {todo.text}
          </li>
        ))}
      </ul>
    </div>
  );
};

export default TodoList;`
            },
            {
                phase: "Code Review",
                ai: "claude",
                task: "Review for accessibility and best practices",
                code: `// Claude's suggestions:
// 1. Add ARIA labels for screen readers
// 2. Implement proper focus management
// 3. Add error handling for edge cases
// 4. Consider using useCallback for performance
// 5. Add TypeScript strict mode compliance`
            },
            {
                phase: "Debugging Session",
                ai: "cursor",
                task: "Interactive debugging and optimization",
                code: `// Cursor helps with:
// - Step-by-step debugging of useState issues
// - Interactive testing of component behavior
// - Real-time code suggestions during development
// - Performance optimization tips`
            }
        ];

        projectFlow.forEach((step, index) => {
            console.log(`\n${index + 1}. ${step.phase} (${step.ai.toUpperCase()}):`);
            console.log(`   Task: ${step.task}`);
            console.log(`   Implementation:`);
            console.log(step.code);
        });
    }

    provideWebDevelopmentNextSteps(): void {
        console.log('\n🚀 Next Steps for Web Developers:');
        console.log('='.repeat(40));

        const steps = [
            "Install Roundtable: pip install roundtable-ai",
            "Set up MCP in your preferred web IDE (VS Code, Cursor)",
            "Try the multi-AI workflow with a simple React component",
            "Experiment with different AIs for frontend vs backend tasks",
            "Build a full-stack project using multiple AI assistants",
            "Create educational content showing AI-assisted web development",
            "Contribute to web development MCP tools and examples"
        ];

        steps.forEach((step, index) => {
            console.log(`${index + 1}. ${step}`);
        });

        console.log('\n📚 Web Development AI Resources:');
        console.log('• React + AI: Learn how to integrate AI assistants in React projects');
        console.log('• Node.js + MCP: Backend development with multiple AI assistants');
        console.log('• Full-Stack AI Workflows: End-to-end project development');
        console.log('• MCP for Web Teams: Collaborative development with unified AI tools');

        console.log('\n🔗 Useful Links:');
        console.log('• Roundtable Documentation: https://github.com/askbudi/roundtable');
        console.log('• MCP Specification: https://modelcontextprotocol.io');
        console.log('• Microsoft Learn: https://learn.microsoft.com');
    }
}

async function main(): Promise<void> {
    const demo = new EducationalMCPDemoTS();

    console.log('🎓 Welcome to the MCP Educational Demo - TypeScript Edition!');
    console.log('This demo shows how Model Context Protocol extends function calling');
    console.log('for modern web development with JavaScript/TypeScript.\n');

    try {
        // Run the demonstration
        await demo.demonstrateMCPConcepts();
        await demo.createWebDevelopmentMCPConfigs();
        await demo.generateWebProjectExample();
        demo.provideWebDevelopmentNextSteps();

        console.log('\n' + '='.repeat(60));
        console.log('🎉 Demo Complete! Ready to start web development with MCP?');
        console.log('💡 Try running: roundtable-ai --check');
        console.log('🌐 Build your next web project with unified AI assistance!');
        console.log('='.repeat(60));

    } catch (error) {
        console.error('❌ Error running demo:', error);
        console.log('💡 Make sure Roundtable is installed: pip install roundtable-ai');
    }
}

// Run the demo if this file is executed directly
if (require.main === module) {
    main().catch(console.error);
}

export { EducationalMCPDemoTS };

File: 11-integrating-with-function-calling/python/mcp-roundtable-example.py
#!/usr/bin/env python3
"""
Model Context Protocol (MCP) Educational Example with Roundtable
================================================================

This example demonstrates how to use the Roundtable MCP Server for
educational purposes, building on the function calling concepts from
the main lesson.

Prerequisites:
- pip install roundtable-ai
- At least one supported AI assistant installed (claude, codex, etc.)

Learning Objectives:
- Understand how MCP extends function calling
- See unified AI assistant management in action
- Learn to compare different AI approaches
- Build educational workflows with multiple AI tools
"""

import subprocess
import json
import os
from typing import Dict, List, Optional

class EducationalMCPDemo:
    """
    Educational demonstration of MCP concepts using Roundtable AI
    """

    def __init__(self):
        self.available_assistants = []
        self.demo_questions = [
            {
                "question": "Explain what a REST API is",
                "category": "concept",
                "best_ai": "gemini"  # Good for explanations
            },
            {
                "question": "Create a simple Python function to calculate factorial",
                "category": "coding",
                "best_ai": "codex"  # Good for clean code
            },
            {
                "question": "Review this code for potential improvements",
                "category": "review",
                "best_ai": "claude"  # Good for code review
            },
            {
                "question": "Show me step-by-step how to debug this function",
                "category": "debugging",
                "best_ai": "cursor"  # Good for interactive help
            }
        ]

    def check_roundtable_installation(self) -> bool:
        """
        Check if Roundtable is properly installed
        """
        try:
            result = subprocess.run(
                ["roundtable-ai", "--help"],
                capture_output=True,
                text=True,
                timeout=10
            )
            return result.returncode == 0
        except (subprocess.TimeoutExpired, FileNotFoundError):
            return False

    def discover_available_assistants(self) -> List[str]:
        """
        Discover which AI assistants are available through Roundtable

        This demonstrates the MCP concept of automatic tool discovery
        """
        try:
            result = subprocess.run(
                ["roundtable-ai", "--check"],
                capture_output=True,
                text=True,
                timeout=30
            )

            if result.returncode == 0:
                # Parse the output to find available assistants
                output_lines = result.stdout.split('\n')
                available = []

                for line in output_lines:
                    if "✓" in line or "available" in line.lower():
                        # Extract assistant name (this is simplified parsing)
                        for assistant in ["claude", "codex", "gemini", "cursor"]:
                            if assistant in line.lower():
                                available.append(assistant)

                return list(set(available))  # Remove duplicates

        except subprocess.TimeoutExpired:
            print("⚠️  Discovery timeout - this is normal for first run")
        except FileNotFoundError:
            print("❌ Roundtable not found. Please install: pip install roundtable-ai")

        return []

    def demonstrate_mcp_concepts(self):
        """
        Educational demonstration of key MCP concepts
        """
        print("🎓 Model Context Protocol (MCP) Educational Demo")
        print("=" * 50)

        # 1. Installation Check
        print("\n1️⃣ Checking Roundtable Installation...")
        if not self.check_roundtable_installation():
            print("❌ Roundtable not installed")
            print("📋 Install with: pip install roundtable-ai")
            return
        print("✅ Roundtable is installed!")

        # 2. Tool Discovery (Core MCP Concept)
        print("\n2️⃣ Discovering Available AI Assistants...")
        self.available_assistants = self.discover_available_assistants()

        if not self.available_assistants:
            print("⚠️  No AI assistants detected")
            print("📋 Install at least one: claude, codex, gemini, or cursor")
            print("📋 Then run: roundtable-ai --check")
        else:
            print(f"✅ Found {len(self.available_assistants)} assistants:")
            for assistant in self.available_assistants:
                print(f"   • {assistant}")

        # 3. Educational Use Cases
        print("\n3️⃣ Educational Use Cases for MCP:")
        self.demonstrate_educational_scenarios()

        # 4. Practical Examples
        print("\n4️⃣ Practical Learning Examples:")
        self.show_learning_examples()

    def demonstrate_educational_scenarios(self):
        """
        Show different educational scenarios where MCP adds value
        """
        scenarios = {
            "Comparative Learning": {
                "description": "Compare how different AIs approach the same problem",
                "example": "Ask 'How do I sort a list?' to multiple AIs and compare approaches"
            },
            "Specialized Tasks": {
                "description": "Use the best AI for each type of learning task",
                "example": "Gemini for concepts, Codex for code, Claude for review"
            },
            "Workflow Integration": {
                "description": "Chain different AIs together for complex projects",
                "example": "Research → Code → Review → Deploy using different AIs"
            },
            "Context Sharing": {
                "description": "All AIs work with the same project context",
                "example": "Switch AIs mid-project without losing context"
            }
        }

        for scenario, details in scenarios.items():
            print(f"\n📚 {scenario}:")
            print(f"   • {details['description']}")
            print(f"   • Example: {details['example']}")

    def show_learning_examples(self):
        """
        Show practical examples students can try
        """
        for i, demo in enumerate(self.demo_questions, 1):
            print(f"\n🎯 Example {i}: {demo['category'].title()} Task")
            print(f"   Question: {demo['question']}")
            print(f"   Best AI for this: {demo['best_ai']}")
            print(f"   Why: {self.get_ai_strength(demo['best_ai'])}")

    def get_ai_strength(self, ai_name: str) -> str:
        """
        Educational explanation of each AI's strengths
        """
        strengths = {
            "claude": "Excellent for code review, explanations, and best practices",
            "codex": "Generates clean, professional code with good structure",
            "gemini": "Great for research, concepts, and detailed explanations",
            "cursor": "Interactive development and step-by-step guidance"
        }
        return strengths.get(ai_name, "General purpose AI assistant")

    def create_sample_mcp_config(self):
        """
        Generate sample MCP configuration for educational environments
        """
        configs = {
            "VS Code (Student IDE)": {
                "file_path": "~/.vscode/settings.json",
                "config": {
                    "mcp.servers": {
                        "roundtable-ai": {
                            "command": "roundtable-ai",
                            "env": {
                                "CLI_MCP_SUBAGENTS": "codex,claude,gemini,cursor",
                                "CLI_MCP_WORKING_DIR": "/path/to/student/project"
                            }
                        }
                    }
                }
            },
            "Claude Desktop (AI Learning)": {
                "file_path": "~/.config/claude_desktop_config.json",
                "config": {
                    "mcpServers": {
                        "roundtable-ai": {
                            "command": "roundtable-ai",
                            "env": {
                                "CLI_MCP_SUBAGENTS": "codex,claude,gemini,cursor"
                            }
                        }
                    }
                }
            }
        }

        print("\n📝 Sample MCP Configurations for Education:")
        print("=" * 45)

        for platform, details in configs.items():
            print(f"\n🔧 {platform}:")
            print(f"   File: {details['file_path']}")
            print(f"   Configuration:")
            print(json.dumps(details['config'], indent=6))

    def provide_next_steps(self):
        """
        Educational guidance for next steps
        """
        print("\n🚀 Next Steps in Your MCP Learning Journey:")
        print("=" * 45)

        steps = [
            "Install Roundtable: pip install roundtable-ai",
            "Check available tools: roundtable-ai --check",
            "Try the basic examples from this demo",
            "Set up MCP in your preferred IDE (VS Code, etc.)",
            "Experiment with different AI assistants for different tasks",
            "Build a learning project using multiple AIs",
            "Explore advanced MCP features and custom tools"
        ]

        for i, step in enumerate(steps, 1):
            print(f"{i}. {step}")

        print("\n📚 Additional Learning Resources:")
        print("• MCP Official Documentation: https://modelcontextprotocol.io")
        print("• Roundtable Repository: https://github.com/askbudi/roundtable")
        print("• Microsoft MCP for Beginners: https://github.com/microsoft/mcp-for-beginners")
        print("• Generative AI Learning Collection: https://aka.ms/genai-collection")

def main():
    """
    Main educational demonstration
    """
    demo = EducationalMCPDemo()

    print("🎓 Welcome to the MCP Educational Demo!")
    print("This demo shows how Model Context Protocol extends function calling")
    print("concepts to create unified AI assistant management for education.\n")

    # Run the full demonstration
    demo.demonstrate_mcp_concepts()
    demo.create_sample_mcp_config()
    demo.provide_next_steps()

    print("\n" + "=" * 60)
    print("🎉 Demo Complete! Ready to start your MCP learning journey?")
    print("💡 Try running: roundtable-ai --check")
    print("=" * 60)

if __name__ == "__main__":
    main()